### PR TITLE
feat: Combine two different stack service implementations

### PR DESCRIPTION
**Pull request by alexlokshin-czi**
We currently have two divergent implementations of stack_service -- one in cli/pkg/stack_service, one in shared/stack; both were identical at some point, and both are being used in CLI through references, but have diverged quite extensively, each one maintaining their own mocks, tests, and tied into the code. This PR will combine both behaviors into a single implementation.

<details open>
  <summary><a href="https://czi-tech.atlassian.net/browse/CCIE-1636" title="CCIE-1636" target="_blank">CCIE-1636</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
  <td>Combine two copies of stack_service into one (cli/pkg/stack_mgr and shared/stack)</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Task
        Task
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Progress</td>
    </tr>
  </table>
</details>

---

Addresses [CCIE-1636]

[CCIE-1636]: https://czi-tech.atlassian.net/browse/CCIE-1636?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: shared/stack/stack_service.go [hunk 1]
```diff
@@ -77,6 +76,12 @@ func (s *StackService) WithWorkspaceRepo(workspaceRepo workspacerepo.WorkspaceRe
 	return s
 }
 
+func (s *StackService) WithContext(env, appName string) *StackService {
```
Comment (jakeyheath) [inline shared/stack/stack_service.go:79]: This function is misleading, because you don't pass in a context object. Why is this moved from the constructor? Both are required arguments for StackService.

Comment (alexlokshin-czi) [reply shared/stack/stack_service.go:79]: One of the implementations has a WithHappyConfig(), which was removed and replaced with the new builder method. I can rename it to WithEnvContext. Passing these arguments into a constructor conflicts with the builder approach, which only see arguments get added over time.

Comment (jakeyheath) [reply shared/stack/stack_service.go:79]: The builder approach still applies, but if the arguments are required, then we should keep them in the constructor. Otherwise you might build a bad object

Comment (alexlokshin-czi) [reply shared/stack/stack_service.go]: Here's a list of args passed through builder calls:

WithBackend
WithExecutor
WithWorkspaceRepo

None of them are present in the constructor args. Creating a constructor with 5 args makes a messy constructor.

Comment (alexlokshin-czi) [reply shared/stack/stack_service.go]: We can convert them to options.

Comment (alexlokshin-czi) [reply shared/stack/stack_service.go]: But I'd prefer to do that in a separate PR.

File: shared/stack/stack.go [hunk 2]
```diff
@@ -312,12 +331,22 @@ func (s *Stack) GetStackInfo(ctx context.Context) (*model.AppStackResponse, erro
...
+	}
+	// TODO: only here until people upgrade their CLIS. remove this in a few weeks
+	metaLegacy := StackMetaLegacy{}
+	err = json.Unmarshal(metaJSON, &metaLegacy)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not unmarshal legacy stack meta")
+	}
+	err = meta.Merge(metaLegacy)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not merge stack meta")
```
Comment (hspitzley-czi) [inline shared/stack/stack.go:349]: I think it's been long enough that we don't need to port this over

File: shared/stack/stack_meta.go [hunk 3]
```diff
@@ -34,6 +37,65 @@ type StackMeta struct {
...
+	Env       string `json:"happy/env"`
+	ImageTag  string `json:"happy/meta/imagetag"`
+	ImageTags string `json:"happy/meta/imagetags"`
+	App       string `json:"happy/app"`
+	Slice     string `json:"happy/meta/slice"`
+	CreatedAt string `json:"happy/meta/created-at"`
+	UpdatedAt string `json:"happy/meta/updated-at"`
+	Owner     string `json:"happy/meta/owner"`
+	Priority  string `json:"happy/meta/priority"`
+}
```
Comment (hspitzley-czi) [inline shared/stack/stack_meta.go:97]: Similarly, I think it's been long enough that we don't need to port this over

File: shared/stack/stack.go [hunk 4]
```diff
@@ -20,6 +20,21 @@ import (
...
+	LastUpdated string            `json:"last_updated,omitempty"`
+	Message     string            `json:"message,omitempty"`
+	Outputs     map[string]string `json:"outputs,omitempty"`
+	Endpoints   map[string]string `json:"endpoints,omitempty"`
+	App         string            `json:"app,omitempty"`
+	GitRepo     string            `json:"git_repo,omitempty"`
+	GitSHA      string            `json:"git_sha,omitempty"`
+	GitBranch   string            `json:"git_branch,omitempty"`
+}
+
```
Comment (hspitzley-czi) [inline shared/stack/stack.go:37]: This appears unused, can it be deleted?

**Review by hspitzley-czi — COMMENTED** [review]
I think this is a dupe of https://github.com/chanzuckerberg/happy/pull/2096 with some additional builder expressions, I'm good with using either implementation but let's clean up some of the old stuff instead of porting it to the new implementation if we go with this PR
